vm_method.c

```diff
@@ -89,33 +128,133 @@ rb_clear_constant_cache(void)
+                struct rb_id_table *mtbl = RCLASS_M_TBL(RCLASS_ORIGIN(owner));
+                rb_id_table_insert(mtbl, mid, (VALUE)new_cme);
+                RB_OBJ_WRITTEN(owner, cme, new_cme);
+                vm_me_invalidate_cache((rb_callable_method_entry_t *)cme);
+
+                RB_DEBUG_COUNTER_INC(cc_invalidate_tree_cme);
+            }
 
-    if (klass == rb_mKernel) {
-	rb_subclass_entry_t *entry = RCLASS_EXT(klass)->subclasses;
+            // invalidate complement tbl
+            if (METHOD_ENTRY_COMPLEMENTED(cme)) {
+                VALUE defined_class = cme->defined_class;
+                struct rb_id_table *cm_tbl = RCLASS_CALLABLE_M_TBL(defined_class);
+                VM_ASSERT(cm_tbl != NULL);
+                int r = rb_id_table_delete(cm_tbl, mid);
+                VM_ASSERT(r == TRUE); (void)r;
+                RB_DEBUG_COUNTER_INC(cc_invalidate_tree_callable);
+            }
 
-	for (; entry != NULL; entry = entry->next) {
-	    struct rb_id_table *table = RCLASS_CALLABLE_M_TBL(entry->klass);
-	    if (table)rb_id_table_clear(table);
-	}
+            RB_DEBUG_COUNTER_INC(cc_invalidate_tree);
+        }
+    }
+}
+
+static void
+clear_iclass_method_cache_by_id(VALUE iclass, VALUE d)
+{
+    VM_ASSERT(RB_TYPE_P(iclass, T_ICLASS));
+    ID mid = (ID)d;
+    clear_method_cache_by_id_in_class(iclass, mid);
+}
+
+static void
+clear_iclass_method_cache_by_id_for_refinements(VALUE klass, VALUE d)
+{
+    if (RB_TYPE_P(klass, T_ICLASS)) {
+        ID mid = (ID)d;
+        clear_method_cache_by_id_in_class(klass, mid);
+    }
+}
+
+void
+rb_clear_method_cache(VALUE klass_or_module, ID mid)
+{
+    if (RB_TYPE_P(klass_or_module, T_MODULE)) {
+        VALUE module = klass_or_module; // alias
+
+        if (FL_TEST(module, RMODULE_IS_REFINEMENT)) {
+            VALUE refined_class = rb_refinement_module_get_refined_class(module);
+            rb_clear_method_cache(refined_class, mid);
+            rb_class_foreach_subclass(refined_class, clear_iclass_method_cache_by_id_for_refinements, mid);
+        }
+        rb_class_foreach_subclass(module, clear_iclass_method_cache_by_id, mid);
+    }
+    else {
+        clear_method_cache_by_id_in_class(klass_or_module, mid);
     }
 }
 
+// gc.c
+void rb_cc_table_free(VALUE klass);
+
+static int
+invalidate_all_cc(void *vstart, void *vend, size_t stride, void *data)
+{
+    VALUE v = (VALUE)vstart;
+    for (; v != (VALUE)vend; v += stride) {
+        if (RBASIC(v)->flags) { // liveness check
+            if (RB_TYPE_P(v, T_CLASS) ||
+                RB_TYPE_P(v, T_ICLASS)) {
+                if (RCLASS_CC_TBL(v)) {
+                    rb_cc_table_free(v);
+                }
+                RCLASS_CC_TBL(v) = NULL;
+            }
+        }
+    }
+    return 0; // continue to iteration
+}
+
+void
+rb_clear_method_cache_all(void)
+{
+    rb_objspace_each_objects(invalidate_all_cc, NULL);
```

Comment:
Is it faster to use `rb_objspace_each_objects` instead of `rb_class_foreach_subclass` here?  I guess this implementation is slower when there are lots of objetcs.  Not tested though.

Reply:
We shouldn't use subclass data structure because we can remove subclasses data structure in future so I use this implementation.

Now it is inefficient implementation, but we can remove it. This function is only used by `using` and `using` can employ smarter invalidation I guess. Also `using` is rarely used method.